### PR TITLE
genericServeT and docs on using a custom monad with generics

### PR DESCRIPTION
**Pull request by k-bx**
Closes https://github.com/haskell-servant/servant/issues/1055

### CODE REVIEW

File: servant-server/src/Servant/Server/Generic.hs [hunk 1]
```diff
@@ -38,6 +41,45 @@ genericServe
...
+  -> Application
+genericServeT f server = serve p $ hoistServer p f (genericServerT server)
+  where
+    p = genericApi (Proxy :: Proxy routes)
+
+-- | Transform a record of routes with custom monad into a WAI 'Application',
+--   while using the given 'Context' to serve the application (contexts are typically
+--   used by auth-related combinators in servant, e.g to hold auth checks) and the given
+--   transformation to map all the handlers back to the 'Handler' monad.
+genericServeT'
```
[the remark this answers is not in the document]
Comment (phadej) [reply servant-server/src/Servant/Server/Generic.hs]: `genericServeTWithContext` is at least the name one would look for. Terrible, but consistent.

[the remark this answers is not in the document]
Comment (phadej) [reply servant-server/src/Servant/Server/Generic.hs]: `genericServeTWithContext` is at least the name one would look for. Terrible, but consistent.

**Comment by k-bx** [comment]
Thanks, addressed your feedback @phadej

File: servant-server/src/Servant/Server/Generic.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alpmestan) [reply servant-server/src/Servant/Server/Generic.hs]: OK, let's go with that name then @k-bx if you don't mind.

**Comment by k-bx** [comment]
I can't find the error in the log https://travis-ci.org/haskell-servant/servant/jobs/444626028 Can anyone take a look please?

**Comment by k-bx** [comment]
Found the error in Raw logs:

```
[;1mGeneric.lhs:121:13: [;1m[31merror:[0m[0m[;1m[0m[0m[;1m
    Not in scope: type constructor or class â€˜ReaderTâ€™
    Perhaps you meant one of these:
      â€˜Headerâ€™ (imported from Servant),
      â€˜Header'â€™ (imported from Servant),
      â€˜Headersâ€™ (imported from Servant)[0m[0m
[;1m[34m    |[0m[0m
[;1m[34m121 |[0m[0m type AppM = [;1m[31mReaderT[0m[0m AppCustomState Handler
```

Not sure why it's happening :/ I've made `import           Control.Monad.Trans.Reader  (ReaderT, ask, runReaderT)` at the top

**Comment by kosmikus** [comment]
@k-bx I cannot see a `Reader`-related import in the file.

**Comment by k-bx** [comment]
Sorry, my bad, confused the file. I'll wrap up the PR now.

**Comment by k-bx** [comment]
8.4.4 builds fine. 8.6 fails in "cabal hell":

```
cabal: Could not resolve dependencies:
[__0] trying: cookbook-basic-auth-0.1 (user goal)
[__1] trying: base-4.12.0.0/installed-4.1... (dependency of
cookbook-basic-auth)
[__2] trying: aeson-1.4.1.0 (dependency of cookbook-basic-auth)
[__3] trying: cookbook-testing-0.0.1 (user goal)
[__4] next goal: servant-quickcheck (dependency of cookbook-testing)
[__4] rejecting: servant-quickcheck-0.0.7.2, servant-quickcheck-0.0.7.0
(conflict: base==4.12.0.0/installed-4.1..., servant-quickcheck => base>=4.8 &&
<4.12)
[__4] rejecting: servant-quickcheck-0.0.6.0, servant-quickcheck-0.0.5.0,
servant-quickcheck-0.0.4.1, servant-quickcheck-0.0.4 (conflict:
base==4.12.0.0/installed-4.1..., servant-quickcheck => base>=4.8 && <4.11)
[__4] rejecting: servant-quickcheck-0.0.3.1, servant-quickcheck-0.0.3.0,
servant-quickcheck-0.0.2.4, servant-quickcheck-0.0.2.3,
servant-quickcheck-0.0.2.2, servant-quickcheck-0.0.2.1 (conflict:
base==4.12.0.0/installed-4.1..., servant-quickcheck => base>=4.8 && <4.10)
[__4] rejecting: servant-quickcheck-0.0.2.0, servant-quickcheck-0.0.1.1
(conflict: aeson==1.4.1.0, servant-quickcheck => aeson>0.8 && <0.12)
[__4] rejecting: servant-quickcheck-0.0.1.0, servant-quickcheck-0.0.0.0
(conflict: aeson==1.4.1.0, servant-quickcheck => aeson>0.10 && <0.12)
[__4] fail (backjumping, conflict set: aeson, base, cookbook-testing,
servant-quickcheck)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: servant-quickcheck, base, aeson,
cookbook-basic-auth, cookbook-testing
```

8.2 and 8.0 fails in Testing.lhs:

```
[;1mTesting.lhs:230:45: [;1m[31merror:[0m[0m[;1m[0m[0m[;1m
    â€¢ Variable not in scope: (<>) :: m0 BaseUrl -> [Char] -> t0
    â€¢ Perhaps you meant one of these:
        data constructor â€˜:>â€™ (imported from Control.Lens),
        â€˜>â€™ (imported from Prelude),
        â€˜<%>â€™ (imported from Servant.QuickCheck)[0m[0m
[;1m[34m    |[0m[0m
[;1m[34m230 |[0m[0m   baseUrl <- parseBaseUrl $ unpack $ esHost [;1m[31m<>[0m[0m ":" <> esPort
[;1m[34m    |[0m[0m[;1m[31m                                             ^^[0m[0m
[0m[0m[0m
[;1mTesting.lhs:230:52: [;1m[31merror:[0m[0m[;1m[0m[0m[;1m
    â€¢ Variable not in scope: (<>) :: t0 -> Text -> IO BaseUrl
    â€¢ Perhaps you meant one of these:
        data constructor â€˜:>â€™ (imported from Control.Lens),
        â€˜>â€™ (imported from Prelude),
        â€˜<%>â€™ (imported from Servant.QuickCheck)[0m[0m
[;1m[34m    |[0m[0m
[;1m[34m230 |[0m[0m   baseUrl <- parseBaseUrl $ unpack $ esHost <> ":" [;1m[31m<>[0m[0m esPort
[;1m[34m    |[0m[0m[;1m[31m                                                    ^^[0m[0m
[0m[0m[0mcabal: Failed to build exe:cookbook-testing from cookbook-testing-0.0.1.
```

I suppose this is on par with master and not related to this PR in particular. So, ready to merge.

**Comment by kosmikus** [comment]
@k-bx I know it's wishful thinking, but it would be great if we could stop calling any dependency problem "cabal hell".

In this case, the problem is that `servant-quickcheck` has an upper bound of `base` that simply makes it incompatible with `ghc-8.6.1`, so there's nothing cabal could do here as long as it's trying to respect that bound. You can try `allow-newer` or use the unreleased/github version which seems to have an updated bound.

**Comment by k-bx** [comment]
@kosmikus aha! Thank you for the explanation. I didn't spend enough time deducing the root cause of the dependency problem from that cabal-install output. I'll refrain from prematurely using the "cabal hell" terminology if that makes fewer people annoyed, cabal (both the lib and the cabal-install) is an amazing piece of engineering that made my life easier so many times.

But my general point stands. I think the boundary problem and the `(<>)` ones should be resolved separately from this PR.

**Comment by phadej** [comment]
@k-bx could you rebase on top of current master, than travis should turn green

**Comment by k-bx** [comment]
Done. Let's see if it builds.

**Pull request merged**